contributors/design-proposals/bulk_watch.md

```diff
@@ -0,0 +1,458 @@
+it has to be possible to extend the implementation without changing the API.
+Spanning multiple resources, resources types or conditions will be more and more
+important for large number of watches. As an example, federation will be adding
+watches for every type it federates. With that in mind, bypassing aggregation
+at the resource type level and going to aggregation over objects with different
+resource types will allow us to more aggresively optimize in the future.
+
+ Moreover, out current REST API doesn't even offer an easy way to handle
+"multiple watches of a given type" within a single request. As a result, instead
+of inventing new top level pattern per type, we can introduce a new resource
+type that follows normal RESTful rules and solves even more generic problem
+of spanning multiple different resource types.
+
+ We will start with introducing a new dedicated API group:
+ ```
+ /apis/bulk.k8s.io/
+ ```
+ that underneath will have a completely separate implementation.
+
+ In this design, we will focus only on bulk list (get) and watch operations.
+Later, we would like to introduce new resources to support bulk create, update
+and delete operations, but that's not part of this design.
+
+ We will start with introducing `list` resource and supporting the following
+two operation:
+```
+POST /apis/bulk.k8s.io/v1/list <body defines filtering>
+POST /apis/bulk.k8s.io/v1/list?watch=1 <body defines filtering>
+```
+ We can't simply make this an http GET request, due to limitations of GET for
+the size (length) of the url (in which we would have to pass filter options).
+This makes the API look similar to our regular API.
+ Note that once we have list fully implemented, adding this kind of watch will
+be faily simple.
+
+ The main drawback of this approach is that it won't allow for dynamic altering
+of watch subscriptions (which we would like to also support).
+As a result, we need another API for watch that will also support incremental
+subscribtions. That's why we will also introduce the following operation:
+```
+websocket /apis/bulk.k8s.io/v1/list?watch=1
+```
+ The more detailed description of this protocol is in the followin subsection.
+
+ For consistency, we may also consider introducing websocket API for handling
+LIST requests, i.e.:
+```
+websocket /apis/bulk.k8s.io/v1/list
+```
+ where first client sends filter definition over the channel and then server
+sends back the response (all further writes from client are ignored). This
+would make the list API consistent with watch.
+
+*TODO: Figure out if we want to proceed with both APIs, just the second or
+maybe some subset of those*
+
+
+### Filtering definition
+
+ We need to make our filtering mechanism to support different resource types at
+the same time. On the other hand, we would like it to be as consistent with all
+other Kubernetes APIs as possible. So we define the selector for bulk operations
+as following:
+
+```
+type BulkSelector struct {
```

Comment:
Will need TypeMeta.

Another name we could use is `AggregateList`.  Or simply `BulkList`.  Or `BulkListOptions`, which is actually pretty close to what we have for `ListOptions` and `DeleteOptions`.

I kind of like `BulkListOptions`.